File: app/src/main/java/org/scp/gymlog/util/SecondTickThread.java
package org.scp.gymlog.util;

import java.util.function.Supplier;

import lombok.Setter;

@Setter
public class SecondTickThread extends Thread {
    private final Supplier<Boolean> onTickListener;
    protected Function onFinishListener;

    public SecondTickThread(Supplier<Boolean> onTickListener) {
        this.onTickListener = onTickListener;
    }

    @Override
    public final void run() {
        try {
            while(onTickListener.get()) Thread.sleep(1000);
        } catch (InterruptedException e) {
            // Thread interrupted
        } finally {
            if (onFinishListener != null) {
                onFinishListener.call();
            }
        }
    }
}

File: app/src/main/java/org/scp/gymlog/util/SecondTickThread.kt
package org.scp.gymlog.util

abstract class SecondTickThread : Thread() {

    open fun onStart() {}
    abstract fun onTick() : Boolean
    open fun onFinish() {}

    override fun run() {
        onStart()
        try {
            while (onTick()) sleep(1000)
        } catch (e: InterruptedException) {
            // Thread interrupted
        } finally {
            onFinish()
        }
    }
}